src/storage/lock_manager/deadlock.rs

```diff
@@ -97,81 +150,86 @@ impl DetectTable {
         let _timer = DETECTOR_HISTOGRAM_VEC.detect.start_coarse_timer();
         TASK_COUNTER_VEC.detect.inc();
 
-        self.now = time::Instant::now_coarse();
-        self.active_expire();
+        let detect_ts = time::Instant::now_coarse();
+        self.active_expire(detect_ts);
 
         // If `txn_ts` is waiting for `lock_ts`, it won't cause deadlock.
-        if self.register_if_existed(txn_ts, lock_ts, lock_hash) {
+        if self.register_if_existed(txn_ts, lock_ts, lock_hash, detect_ts) {
             return None;
         }
 
-        if let Some(deadlock_key_hash) = self.do_detect(txn_ts, lock_ts) {
-            return Some(deadlock_key_hash);
+        let dead_lock = self.do_detect(txn_ts, lock_ts, detect_ts);
+        if dead_lock.is_none() {
+            self.register(txn_ts, lock_ts, lock_hash, detect_ts);
         }
-        self.register(txn_ts, lock_ts, lock_hash);
-        None
+        dead_lock
     }
 
     /// Checks if there is a edge from `wait_for_ts` to `txn_ts`.
-    fn do_detect(&mut self, txn_ts: u64, wait_for_ts: u64) -> Option<u64> {
-        let now = self.now;
+    fn do_detect(
+        &mut self,
+        txn_ts: u64,
+        wait_for_ts: u64,
+        detect_ts: time::Instant,
+    ) -> Option<u64> {
         let ttl = self.ttl;
-
         let mut stack = vec![wait_for_ts];
-        // Memorize the pushed vertexes to avoid duplicate search.
-        let mut pushed: HashSet<u64> = HashSet::default();
-        pushed.insert(wait_for_ts);
+        // visited collected the txn which is already visited by BFS.
+        let mut visited: HashSet<u64> = HashSet::default();
+        visited.insert(wait_for_ts);
         while let Some(wait_for_ts) = stack.pop() {
-            if let Some(wait_for) = self.wait_for_map.get_mut(&wait_for_ts) {
-                // Remove expired edges.
-                wait_for.retain(|_, locks| !locks.is_expired(now, ttl));
-                if wait_for.is_empty() {
-                    self.wait_for_map.remove(&wait_for_ts);
-                } else {
-                    for (lock_ts, locks) in wait_for {
-                        if *lock_ts == txn_ts {
-                            return Some(locks.hashes[0]);
-                        }
-                        if !pushed.contains(lock_ts) {
-                            stack.push(*lock_ts);
-                            pushed.insert(*lock_ts);
-                        }
-                    }
+            let txns = if let Some(wait_for) = self.wait_for_map.get_mut(&wait_for_ts) {
+                wait_for
+            } else {
+                continue;
+            };
+
+            // Remove expired edges.
+            let is_empty = txns.clean_expired_txns(detect_ts, ttl);
+            if is_empty {
+                self.wait_for_map.remove(&wait_for_ts);
+                continue;
+            }
+            // do BFS
```

Comment:
Strictly speaking, it doesn't seems to be a BFS. I think it's still a DFS, but just removed recursion.